controller/src/main/java/org/jboss/as/controller/OperationContextImpl.java

```diff
@@ -300,14 +310,23 @@ private boolean validateCapabilities() {
                         // Problem wasn't a capability removal.
                         // See what step(s) added this requirement
                         Set<Step> bereft = addedRequirements.get(reqReq);
-                        assert bereft != null && bereft.size() > 0;
-                        for (Step step : bereft) {
-                            Set<CapabilityId> set = missingForStep.get(step);
-                            if (set == null) {
-                                set = new HashSet<>();
-                                missingForStep.put(step, set);
+                        if (bereft != null && bereft.size() > 0) {
+                            for (Step step : bereft) {
+                                Set<CapabilityId> set = missingForStep.get(step);
+                                if (set == null) {
+                                    set = new HashSet<>();
+                                    missingForStep.put(step, set);
+                                }
+                                set.add(required);
                             }
-                            set.add(required);
+                        } else {
+                            if (unexplainedProblem == null) {
+                                unexplainedProblem = new StringBuilder(ControllerLogger.ROOT_LOGGER.requiredCapabilityMissing());
+                            }
+                            String formattedCapability = ignoreContext
+                                    ? ControllerLogger.ROOT_LOGGER.formattedCapabilityName(reqReq.getRequiredName())
+                                    : ControllerLogger.ROOT_LOGGER.formattedCapabilityId(reqReq.getRequiredName(), reqReq.getDependentContext().getName());
+                            unexplainedProblem = unexplainedProblem.append('\n').append(formattedCapability);
```

Comment:
We should probably use `append(System.lineSeparator())` so Windows will get a nice looking message as well.